lib/parser.js

```diff
@@ -1,41 +1,171 @@
+  parcelId,           // id, // parcelId: 4b
+  parcelType,         // type, // parcelType: 1b
+  compression,  // compression, // 1b: no = 0, gzip = 1
+  encoding,     // encoding,
+  // 1b: binary = 0, jstp = 1, json = 2, bson = 3, v8 = 4
+  length        // length //string !!! // 8b
+}) => {
+  const parcel = Buffer.alloc(PARCEL_LENGTH);
+  parcel.writeInt8(constObj.STRUCT_PARCEL);
+  parcel.writeInt32LE(parcelId, 1);
+  parcel.writeInt8(parcelType, 5);
+  parcel.writeInt8(compression, 6);
+  parcel.writeInt8(encoding, 7);
+  length = '0'.repeat(8 - length.length) + length;
+  parcel.write(length, 8);
+  return parcel;
+};
+
+const readChunk = (buffer, length) => {
   // Read chunkk from buffer
   // Return chunk object
-  const rest = Buffer.from(buffer, 1, CHUNK_LENGTH);
-  return rest;
+  const chunk = { structType: constObj.STRUCT_CHUNK };
+  chunk.parcelId = buffer.readInt32LE(1);
+  chunk.chunkId = buffer.readInt32LE(5);
+  chunk.flags = buffer.readInt8(9);
+  chunk.length = buffer.readInt16LE(10);
+  chunk.payload = buffer.slice(12, length);
+  return chunk;
 };
 
-const handshake = (buffer) => {
+const readPayloadLength = (buffer) =>
+  buffer.readInt16LE(10);
+
+const partPayload = (string, size = 2048) => {
+  // Turn data into an array of payloads
+  let chunkId = 1;
+  let payload = string.substring(0, size);
+  string = string.slice(size);
+
+  const arrayOfPayloads = [
+    {
+      chunkId: chunkId++,
+      payload: Buffer.from(payload),
+      length: payload.length
+    }
+  ];
+
+  while (string.length > 0) {
+    payload = string.substring(0, size);
+    string = string.slice(size);
+    arrayOfPayloads.push({
+      chunkId: chunkId++,
+      payload: Buffer.from(payload),
+      length: payload.length
+    });
+  }
+
+  return arrayOfPayloads;
+};
+
+const chunk = ({
+  // Encode chunk structure to Buffer
+  parcelId, // parcelId, // 4b
+  chunkId,  // chunkId, // 4b
+  flag,    // flags, // 1b: more = 1, stop = 2, pause = 4, resume = 8
+  length,   // length, // 2b
+  payload   // payload // Buffer
+}) => {
+  const chunk = Buffer.alloc(CHUNK_LENGTH);
+  chunk.writeInt8(constObj.STRUCT_CHUNK);
+  // Write chunk fields to buffer
+  chunk.writeInt32LE(parcelId, 1);
+  chunk.writeInt32LE(chunkId, 5);
+  chunk.writeInt8(flag, 9);
+  chunk.writeInt16LE(length, 10);
+  return Buffer.concat([chunk, payload], CHUNK_LENGTH + length);
+};
+
+const readHandshake = (buffer) => {
   // Read handshake from buffer
   // Return handshake object
   const rest = Buffer.from(buffer, 0, HANDSHAKE_LENGTH);
-  return rest;
+  const handshake = {};
+  handshake.version = rest.readInt16LE(0);
+  handshake.status = rest.readInt8(2);
+  handshake.reserved = rest.readInt8(3);
+  handshake.token = rest.readInt32LE(4);
+  return handshake;
+};
+
+const handshake = ({
+  // Encode handshake structure to Buffer
+  status,   // status, // 1b: 0 = new, 1 = restore
+  reserved, // reserved, // 1b
+  token // token // 32b (optional)
```

Comment:
Set the default value of token as `''`, because `Buffer.write` requires its argument to be a string.